relays/messages/src/relay_strategy/enforcement_strategy.rs

```diff
@@ -0,0 +1,242 @@
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! enforcement strategy
+
+use num_traits::{SaturatingAdd, Zero};
+
+use bp_messages::{MessageNonce, Weight};
+use bp_runtime::messages::DispatchFeePayment;
+
+use crate::{
+	message_lane::MessageLane,
+	message_lane_loop::{
+		MessageDetails, RelayerMode, SourceClient as MessageLaneSourceClient,
+		TargetClient as MessageLaneTargetClient,
+	},
+	message_race_loop::NoncesRange,
+	relay_strategy::{RelayMessagesBatchReference, RelayReference, RelayStrategy},
+};
+
+/// Do hard check and run soft check strategy
+#[derive(Clone)]
+pub struct EnforcementStrategy<Strategy: RelayStrategy> {
+	strategy: Strategy,
+}
+
+impl<Strategy: RelayStrategy> EnforcementStrategy<Strategy> {
+	pub fn new(strategy: Strategy) -> Self {
+		Self { strategy }
+	}
+}
+
+impl<Strategy: RelayStrategy> EnforcementStrategy<Strategy> {
+	pub async fn decide<
+		P: MessageLane,
+		SourceClient: MessageLaneSourceClient<P>,
+		TargetClient: MessageLaneTargetClient<P>,
+	>(
+		&self,
+		reference: RelayMessagesBatchReference<P, SourceClient, TargetClient>,
+	) -> Option<MessageNonce> {
+		let mut hard_selected_count = 0;
+		let mut soft_selected_count = 0;
+
+		let mut selected_weight: Weight = 0;
+		let mut selected_count: MessageNonce = 0;
+
+		let hard_selected_begin_nonce =
+			reference.nonces_queue[reference.nonces_queue_range.start].1.begin();
+
+		// relay reference
+		let mut relay_reference = RelayReference {
+			lane_source_client: reference.lane_source_client.clone(),
+			lane_target_client: reference.lane_target_client.clone(),
+
+			selected_reward: P::SourceChainBalance::zero(),
+			selected_cost: P::SourceChainBalance::zero(),
+			selected_size: 0,
+
+			total_reward: P::SourceChainBalance::zero(),
+			total_confirmations_cost: P::SourceChainBalance::zero(),
+			total_cost: P::SourceChainBalance::zero(),
+
+			hard_selected_begin_nonce,
+			selected_prepaid_nonces: 0,
+			selected_unpaid_weight: 0,
+
+			index: 0,
+			header_id: Default::default(),
+			nonce: 0,
+			details: MessageDetails {
+				dispatch_weight: 0,
+				size: 0,
+				reward: P::SourceChainBalance::zero(),
+				dispatch_fee_payment: DispatchFeePayment::AtSourceChain,
+			},
+		};
+
+		let all_ready_nonces = reference
+			.nonces_queue
+			.range(reference.nonces_queue_range.clone())
+			.map(|(header_id, ready_nonces)| (header_id, ready_nonces.iter()))
+			.enumerate();
+
+		for (index, (header_id, nonces)) in all_ready_nonces {
+			relay_reference.index = index;
+			relay_reference.header_id = header_id.clone();
```

Comment:
We need to use the block number to determine whether the relay has timed out, so give up using flat_map and put the block number in the reference.